Enable a sequencer to re-sequence batches

cffls opened the pull request:
Feature PR for https://github.com/0xPolygonHermez/cdk-erigon/issues/667

Overall workflow:

- Stop sequencer
- Unwind to target batch (Using command under https://github.com/0xPolygonHermez/cdk-erigon/blob/zkevm/cmd/integration/commands/stage_stages_zkevm.go). e.g. `CDK_ERIGON_SEQUENCER=1 go run ./cmd/integration state_stages_zkevm --config=/path/to/config.yaml --unwind-batch-no=8 --chain dynamic-kurtosis --datadir /path/to/chain/data`
- Backup data stream files in case the files are corrupted during re-sequencing
- Restart sequencer
- The sequencer will detect rollback and start re-sequencing automatically
- On re-sequencing, L2 blocks will either be the same as before, split into multiple, but they can never be merged. Some properties of rollback behavior:

  -   If a block is split into multiple blocks after re-sequencing, the split blocks will have the same timestamp as the original block.
  -   A new L2 block won’t include transactions that belonged to different blocks that got rolled back.
  -   A new sequenced batch won’t include blocks that belonged to different batches that got rolled back.
  
  
 **Tested the following rollback scenarios in kurtosis:**
 The original chain (before resequenced) contains both empty blocks and non-empty blocks with uniswap transactions.
 
 1. Condition: `ResequenceStrict=true`. No L1InfoUpdate on L1. No counter limit change.
  Result: Resequenced L2 blocks have exactly the same block hashes as previously created. 
 2. Condition: `ResequenceStrict=true`. No L1InfoUpdate on L1. Lowered counter limit. 
  Result: An error was raised because a previous batch had to be split into two batches.
 3. Condition: `ResequenceStrict=false`. No L1InfoUpdate on L1. Lowered counter limit. 
 Result: Resequence completed with more number of batches than before.
 4. Condition: `ResequenceStrict=true`. Multiple L1InfoUpdates on L1. No counter limit change.
 Result: Same number of blocks and batches were resequenced. L2 block hash were the same.
 5. Condition: `ResequenceStrict=false`. Multiple L1InfoUpdates on L1. No counter limit change.
 Result: Same number of blocks and batches were resequenced. L2 block hash were different because different L1InfoUpdate index were used.

cffls committed:
Enable a sequencer to re-sequence batches

Stefan-Ethernal commented on cmd/utils/flags.go:
```diff
@@ -481,6 +481,11 @@ var (
 		Value: 0,
 	}
+	ResequenceStrictMode = cli.BoolFlag{
+		Name:  "zkevm.resequence-strict",
```
WDYT?
```suggestion
		Name:  "zkevm.sequencer-resequence-strict",
```

cffls replied:
Agreed. It would make the naming more consistent.

Stefan-Ethernal commented on zk/datastream/client/stream_client.go:
```diff
@@ -357,6 +357,43 @@ LOOP:
+	batches := make([][]*types.FullL2Block, end-start+1)
+
+	for i := start; i <= end; i++ {
+		for {
```
Just to check if my understanding is correct here: data streamer tracks the last fetched stream entry (so no need to move the bookmark/iterator to read the next entry, as it is handled internally)?

And here basically for a single batch (inner for loop), we are reading entries (multiple blocks within the single batch) until we hit the BATCH_END, meaning we done with the current batch and fetching the next one (outer for loop).

cffls replied:
Yes, it is handled internally. The entry is iterated after a bookmark.

Stefan-Ethernal commented on zk/datastream/client/stream_client.go:
```diff
@@ -357,6 +357,43 @@ LOOP:
+				} else {
+					batches[block.BatchNumber] = append(batches[block.BatchNumber], block)
+				}
+			}
```
Instead of checking whether the slice is initialized for certain batch number each time a block is retrieved, we could probably do that only when `batchStart` entry is retrieved from the data stream (slightly more readable IMO). Hopefully there would not be empty batches, because in that case, your approach is better (because it instantiates slice more lazily). WDYT?

```suggestion
			block, batchStart, batchEnd, _, _, _, err := c.readFullBlockProto()
			if err != nil {
				return nil, err
			}

			if batchEnd != nil && batchEnd.Number == end {
				break
			}

			if batchStart != nil {
				batches[block.BatchNumber] = []*types.FullL2Block{}
			}

			if block != nil {
				batches[block.BatchNumber] = append(batches[block.BatchNumber], block)
			}
```

Stefan-Ethernal commented on zk/tx/tx.go:
```diff
@@ -189,6 +189,8 @@ func DecodeBatchL2Blocks(txsData []byte, forkID uint64) ([]DecodedBatchL2Data, e
 
+type TxDecoder func([]byte, byte, uint64) (types.Transaction, uint8, error)
+
 func DecodeTx(encodedTx []byte, efficiencyPercentage byte, forkId uint64) (types.Transaction, uint8, error) {
```
I'm aware that this is the legacy code, but as a side note:
it's strange to me that `efficiencyPercentage` parameter is behaving as input & output parameter. I don't see the point. :confused:

```suggestion
func DecodeTx(encodedTx []byte, efficiencyPercentage byte, forkId uint64) (types.Transaction, uint8, error) {
```

cffls replied:
I think it is used to encode the transaction, but agreed it may be necessary as part of the output. I will leave it as it is for now.

Stefan-Ethernal commented on zk/tx/tx.go:
```diff
@@ -189,6 +189,8 @@ func DecodeBatchL2Blocks(txsData []byte, forkID uint64) ([]DecodedBatchL2Data, e
 }
 
+type TxDecoder func([]byte, byte, uint64) (types.Transaction, uint8, error)
+
```
Consider using named parameters, for the sake of brevity.

```suggestion
type TxDecoder func(encodedTx []byte, gasPricePercentage uint8, forkID uint64) (types.Transaction, uint8, error)
```

Stefan-Ethernal commented on zk/datastream/client/stream_client.go:
```diff
@@ -357,6 +357,43 @@ LOOP:
 	return err
 }
 
+func (c *StreamClient) ReadBatches(start uint64, end uint64) ([][]*types.FullL2Block, error) {
```
Optional consideration: should we expose the function to the [interface](https://github.com/cffls/erigon/blob/5f9d291b33150793a9af43330c6caf59454430a3/zk/stages/stage_batches.go#L74)?

cffls replied:
Yeah it is a good idea. Will do.

Stefan-Ethernal commented on zk/stages/stage_sequence_execute_utils.go:
```diff
@@ -498,3 +499,69 @@ func (bdc *BlockDataChecker) AddTransactionData(txL2Data []byte) bool {
+
+func (r *ResequenceBatchJob) UpdateLastProcessedTx(h common.Hash) {
+	if idx, ok := r.txIndexMap[h]; ok {
+		block := r.batchToProcess[idx[0]]
```
Consider capturing idx[0] and idx[1] to some local vars, as they are used on several places (both for readability and slight performance gain). WDYT?

cffls replied:
Will update the integer slice to a struct.

Stefan-Ethernal commented on zk/stages/stage_sequence_execute_utils.go:
```diff
@@ -498,3 +499,69 @@ func (bdc *BlockDataChecker) AddTransactionData(txL2Data []byte) bool {
 	return false
 }
+
+type ResequenceBatchJob struct {
```
Just a consideration, if you like it better, but doesn't really matter.

```suggestion
type ResequenceBatchTracker struct {
```

cffls replied:
I'd prefer to use Job, because it can represent both the job and the job progress :D

Stefan-Ethernal commented on zk/stages/stage_sequence_execute_utils.go:
```diff
@@ -498,3 +499,69 @@ func (bdc *BlockDataChecker) AddTransactionData(txL2Data []byte) bool {
+	StartBlockIndex int
+	StartTxIndex    int
+	txIndexMap      map[common.Hash][]int
+}
```
WDYT about such struct (instead of having slice with two ints)?

```suggestion
	txIndexMap      map[common.Hash]*txMetadata
}

type txMetadata struct{
	blockNum int
	txIndex int
}
```

cffls replied:
Good suggestion. Will change accordingly.

Stefan-Ethernal commented on zk/stages/stage_sequence_execute_utils.go:
```diff
@@ -498,3 +499,69 @@ func (bdc *BlockDataChecker) AddTransactionData(txL2Data []byte) bool {
+			r.txIndexMap[tx.Hash()] = []int{r.StartBlockIndex, i}
+			if err != nil {
+				return nil, fmt.Errorf("decode tx error: %v", err)
+			}
```
Swap these two lines?
```suggestion
            if err != nil {
				return nil, fmt.Errorf("decode tx error: %v", err)
			}
			r.txIndexMap[tx.Hash()] = []int{r.StartBlockIndex, i}
```

Stefan-Ethernal commented on zk/stages/stage_sequence_execute.go:
```diff
@@ -47,6 +44,83 @@ func SpawnSequencingStage(
+		if err != nil {
+			return err
+		}
+		dsClient := client.NewClient(ctx, fmt.Sprintf("%s:%d", cfg.zk.DataStreamHost, cfg.zk.DataStreamPort), cfg.zk.DatastreamVersion, cfg.zk.L2DataStreamerTimeout, uint16(latestForkId))
```
Would it make sense putting `dsClient` to the `SequenceBlockCfg` (https://github.com/cffls/erigon/blob/5f9d291b33150793a9af43330c6caf59454430a3/zk/stages/stage_sequence_execute_utils.go#L69) and making it "singleton" (it is probably already instantiated from the outside)?

Similar was done for batches stage (https://github.com/cffls/erigon/blob/5f9d291b33150793a9af43330c6caf59454430a3/zk/stages/stage_batches.go#L89).

cffls replied:
I am not sure if the data stream client is initiated in the sequencer, because the sequencer in most cases will only have the data stream server. I think it is okay to only create the client when it is needed when resequencing.

Stefan-Ethernal commented on zk/stages/stage_sequence_execute.go:
```diff
@@ -47,6 +44,83 @@ func SpawnSequencingStage(
+			}
+		}
+
+		cfg.datastreamServer.UnwindToBatchStart(lastBatch)
```
Handle the error

cffls replied:
Done.

Stefan-Ethernal commented on zk/stages/stage_sequence_execute.go:
```diff
@@ -47,6 +44,83 @@ func SpawnSequencingStage(
+		return err
+	}
+
+	if lastBatch >= highestBatchInDS {
```
Is this check enough or it is possible to make it even more strict?

Not sure does if this makes sense, but for example:
- if lastBatch == highestBatchInDS + 1
- comparing parent hashes etc.

cffls replied:
Good question. I think the check will be sufficient for now, since the only case where the inconsistency can happen is when the sequencer is unwinded.

Stefan-Ethernal commented on zk/stages/stage_sequence_execute.go:
```diff
@@ -47,6 +44,83 @@ func SpawnSequencingStage(
+			return err
+		}
+		dsClient := client.NewClient(ctx, fmt.Sprintf("%s:%d", cfg.zk.DataStreamHost, cfg.zk.DataStreamPort), cfg.zk.DatastreamVersion, cfg.zk.L2DataStreamerTimeout, uint16(latestForkId))
+
```
Do we need to call dsClient.Start(), before invoking reading batches?

P.s. this comment can be ignored if we use the "singleton" dsClient instance if I'm not mistaken.

cffls replied:
Good catch! We will probably need it :D

Stefan-Ethernal commented on zk/stages/stage_sequence_execute.go:
```diff
@@ -47,6 +44,83 @@ func SpawnSequencingStage(
+		}
+		dsClient := client.NewClient(ctx, fmt.Sprintf("%s:%d", cfg.zk.DataStreamHost, cfg.zk.DataStreamPort), cfg.zk.DatastreamVersion, cfg.zk.L2DataStreamerTimeout, uint16(latestForkId))
+
+		batches, err := dsClient.ReadBatches(lastBatch+1, highestBatchInDS)
```
Correct me if I'm wrong, but if we have reached this point, it means that `highestBatchInDS < lastBatch` (negation of this condition https://github.com/cffls/erigon/blob/5f9d291b33150793a9af43330c6caf59454430a3/zk/stages/stage_sequence_execute.go#L57), so it doesn't seem like the interval for fetching batches is correct.

cffls replied:
I think when we reach this point, `lastBatch` will be smaller than `highestBatchInDS`, but will double check.

Stefan-Ethernal replied:
Yes, you are right, my bad!

Stefan-Ethernal reviewed (commented):
Leaving some early feedback.

hexoscott commented:
Hey @cffls a few thoughts on the changes here:
- the sequencer is the master source of the datastream on the network so connecting to itself using a stream client to get the batches to resequence might not be a good approach.
- the sequencer code already has quite a few branches in logic around limbo and l1 recovery so the 3rd `resequence` might get confusing.

To my knowledge the process for marking a batch as bad would be along the lines of:
- restart sequencer with a flag to mark an explicit stop batch height
- allow it to run to this height where it will pause
- let the sequence sender etc. catch up and get the network in a verified state
- stop the sequencer
- make the TX to the L1 to mark a batch as bad
- wait for the block including this TX to be finalised on the L1
- start the sequencer in L1 recovery mode...

This last step in my head would start the code we already have for L1 recovery but some new L1 tracking code would be scanning the L1 for this event around a bad batch and storing it in the database.  Once we know the batch is bad we can immediately roll back the datastream to the last good batch along with the database using the normal unwind feature, from there the normal L1 recovery code would start but as part of our "bad batch" checks during recovery we'd have one more check for this scenario, when we encounter it we simply mark the batch as bad as do already and move on to normal recovery.  This way blocks and batches would be an exact replicate of the L1 data apart from skipping the bad batch.  Once L1 recovery is completed to the previous tip you can restart the sequencer as normal and the network continues.

I think this approach works out simpler and we don't need to do anything fancy around splitting blocks or handling timestamps, we just recreate the L1 exactly as it was before.  This has the advantage of also handling things like parent hashes automatically without fuss.

What do you think?

hexoscott commented:
I think maybe there has been something missed here with regards to the banana fork requirements of the batch that will be removed being something that happens on the L1 to affect the chains history.

hexoscott reviewed (changes requested):
Hi @cffls - after the discussion yesterday and looking over the changes in a new light it's looking good and making a lot more sense approach wise to me now.

There are a couple of things that would help here I think:
1. Making this process really explicit by using a flag, if it isn't set then the sequencer won't perform the checks to see if it is behind and just behave as normal.
2. Rather than using a datastream client pointing back to itself, we have the `data_stream_server.go` file that should let you work with the stream without needing the client and it will work directly with the files on disk.  That approach seems cleaner to me.

If we can make those changes I think we'll be in a safe place to test this and get it merged.

cffls commented:
> There are a couple of things that would help here I think:
Making this process really explicit by using a flag, if it isn't set then the sequencer won't perform the checks to see if it is behind and just behave as normal.
Rather than using a datastream client pointing back to itself, we have the data_stream_server.go file that should let you work with the stream without needing the client and it will work directly with the files on disk. That approach seems cleaner to me.

Thanks @hexoscott . I've addressed your feedbacks accordingly.

hexoscott commented:
Great 😄  could you give me a nudge once the latest zkevm branch stuff is merged in so I don't miss it 🙏

cffls commented:
Thanks @hexoscott . I've merged from `zkevm` branch and resolved the conflicts.

hexoscott commented on zk/stages/stage_sequence_execute.go:
```diff
@@ -25,6 +26,72 @@ func SpawnSequencingStage(
+		return err
+	}
+
+	if !cfg.zk.SequencerResequence || lastBatch >= highestBatchInDS {
```
I think we run the risk here of the sequencer just continuing on after resequencing is complete with this logic.  Ideally we want a hard stop once resequencing is complete and the node will need restarting without the resequence flag being set.

cffls replied:
Thanks, this is a great point! I've added a check at the beginning and the node will stop sequencing new blocks after re-sequencing completes.

hexoscott reviewed (changes requested):
Nice updates 😄 just one little thing around the use of the flag in the sequencing stage

cffls committed:
Stop producing blocks after resequencing completes

cffls committed:
Introduce flag reuseL1InfoIndex

cffls committed:
Merge remote-tracking branch 'origin/zkevm' into resequence

cffls committed:
Merge remote-tracking branch 'origin/zkevm' into resequence

cffls committed:
Merge remote-tracking branch 'origin/zkevm' into resequence

cffls committed:
Merge remote-tracking branch 'origin/zkevm' into resequence

cffls committed:
Merge remote-tracking branch 'origin/zkevm' into resequence

cffls committed:
Merge remote-tracking branch 'origin/zkevm' into resequence

cffls committed:
Move ReadParsedProto back to data stream client in order to reduce merge conflict

cffls committed:
Merge remote-tracking branch 'origin/zkevm' into resequence

cffls committed:
Skip iteration when the last batch is incomplete

cffls committed:
Merge remote-tracking branch 'origin/zkevm' into resequence

cffls committed:
Merge remote-tracking branch 'origin/zkevm' into resequence

cffls committed:
Merge remote-tracking branch 'origin/zkevm' into resequence

cffls committed:
Merge remote-tracking branch 'origin/zkevm' into resequence

cffls committed:
Wait for pending verifications during resequencing

cffls committed:
Merge remote-tracking branch 'origin/zkevm' into resequence

cffls committed:
Merge remote-tracking branch 'origin/zkevm' into resequence

hexoscott commented on cmd/integration/commands/stage_stages_zkevm.go:
```diff
@@ -101,6 +95,8 @@ func unwindZk(ctx context.Context, db kv.RwDB) error {
 		return err
 	}
 
+	stages.SaveStageProgress(tx, stages.HighestSeenBatchNumber, unwindBatchNo)
```
I think this would make more sense in the stage responsible for writing it.  i.e. in stage batches we handle it in the unwind there.  Maybe this needs an unwind in the sequencer related stages?

cffls replied:
Good point, and I think this is unnecessary now, because it will be covered by unwind in sequencing stage [here](https://github.com/0xPolygonHermez/cdk-erigon/blob/f911a0daf793b77fa03da3b4f57838630ffe42c6/zk/stages/stage_sequence_execute_unwind.go#L64).

hexoscott commented on zk/stages/stage_sequence_execute.go:
```diff
@@ -28,6 +29,78 @@ func SpawnSequencingStage(
+			return err
+		}
+
+		err = cfg.datastreamServer.UnwindToBatchStart(lastBatch + 1)
```
we might want to save this part until the state has been handled as once this has been done there is no recovering it if there is an error handling the re-sequencing elsewhere

cffls replied:
In order to write to datastream only after everything is completed, a big refactoring seems required, because the datastream is currently updated in this function [updateStreamAndCheckRollback](https://github.com/0xPolygonHermez/cdk-erigon/blob/7ce4998778d9ba29ae9c0fe9a71f9414863b62ee/zk/stages/stage_sequence_execute.go#L518). 

As mentioned in the description of this PR, a solution to avoid losing old data stream is to backup the data stream files before resequence:
> Backup data stream files in case the files are corrupted during re-sequencing

hexoscott replied:
OK, considering this resequencing is really not going to be a common thing that makes sense that there are some manual steps.  It would also be recoverable from an RPC node so long as it was synced at the time.

hexoscott commented on zk/stages/stage_sequence_execute_utils.go:
```diff
@@ -282,6 +297,10 @@ func prepareL1AndInfoTreeRelatedStuff(sdb *stageDb, batchState *BatchState, prop
 
+	if batchState.isResequence() && l1TreeUpdateIndex == 0 {
+		shouldWriteGerToContract = false
+	}
```
I don't think we need this, if there is no change in the l1 info tree index we won't write the GER anyway.

cffls replied:
Agreed. Will remove.

hexoscott commented on zk/stages/stage_sequence_execute_utils.go:
```diff
@@ -489,3 +508,78 @@ func (bdc *BlockDataChecker) AddTransactionData(txL2Data []byte) bool {
+
+func NewResequenceBatchJob(batch []*dsTypes.FullL2Block) *ResequenceBatchJob {
+	return &ResequenceBatchJob{
+		batchToProcess:  batch,
```
maybe a misleading name?  blocksToProcess perhaps?  Similar for the param name being batch.  Typically this means a batch number rather than a collection of blocks.

cffls replied:
This struct represent the original batch that might be preserved or split after resequencing. The name seems to match to what it contains.

hexoscott reviewed (changes requested):
Code looks good, just a few comments from the last time I had a peek at it.

cffls committed:
Address PR comments

cffls committed:
Merge branch 'zkevm' into resequence
